Correctly identify bitfield constants as CONSTANT values

chrisr-diffblue opened the pull request:
Previously constant bitfield values were identified as TWO_VALUE by
default, meaning they became TOP immediately. This change makes them
be correctly identified as constants and hence have their values tracked
appropriately.

chrisr-diffblue commented on src/analyses/variable-sensitivity/variable_sensitivity_object_factory.cpp:
```diff
@@ -126,7 +127,7 @@ abstract_object_pointert variable_sensitivity_object_factoryt::
       followed_type, top, bottom, e, environment, ns);
   default:
-    assert(false);
+    UNREACHABLE;
```
This was a bit of drive-by cleanup to keep the Linter happy...

peterschrammel replied:
Cleanups should go into a separate commit.

peterschrammel reviewed (commented):
Maybe add a test that shows the effect of this change.

martin-cs commented:
Including the test case they sent would be nice but I'm happy this is the right code in the right place.

thk123 reviewed (changes requested):
As already pointed out this should definitely have a test that previously didn't work and now does. Otherwise looks good

chrisr-diffblue committed:
Correctly identify bitfield constants as CONSTANT values

Previously constant bitfield values were identified as TWO_VALUE by
default, meaning they became TOP immediately. This change makes them
be correctly identified as constants and hence have their values tracked
appropriately.

chrisr-diffblue commented:
Thanks for the comments folks - dumb mistake on my part not to include a regression test, now added. Its a simplified version of the original customer test case, but includes the important parts. Without the fix in this PR, the value of the `.b0` field will be TOP.

I've separated the Lint cleanup into a separate commit too, but still included in this PR to stop the Cpp Linter complaining.

thk123 commented on regression/goto-analyzer/variable-sensitivity-bit-field-constants/test.desc:
```diff
@@ -0,0 +1,7 @@
+--function main --variable --structs
+^EXIT=0$
+^SIGNAL=0$
+^bs.*\(\).*->.*\{\.byte=00001010.*@.*,.*\.b0=1.*@.*\}.*$
```
Non-blocking, but you may prefer to:

```
__CPROVER_assert(bs.bo==1);
```

Then run with `--verify` and check for:

```
^\[main.assertion.1\] .* assertion bs.bo==1: Success$
```

As a slightly less fiddly way to validate this.

martin-cs replied:
@thk123 you beat me to it!  Yes, that is a better way of recognising precision at a given point.

chrisr-diffblue replied:
Ah, cool - I like that, thanks. A bit more knowledge squirrelled away for future reference...

chrisr-diffblue committed:
Driveby cleanup to keep Lint happy

chrisr-diffblue commented:
Regression test updated to follow @thk123 suggestion.

martin-cs commented:
Test case failures are in function sensitivity which I'm assured (@thk123 ?) are expected with the current branch and in the windows build, which, again, is not expected to work right now.  Thus I'm happy to merge this.
